src/commonMain/kotlin/org/kosat/CDCL.kt

```diff
@@ -0,0 +1,274 @@
+package org.kosat
+
+import kotlin.math.abs
+
+//CDCL
+fun solveCnf(cnf: CnfRequest): List<Int>? {
+    val clauses = ArrayList(cnf.clauses.map { ArrayList(it.lit.toList()) })
+    return CDCL(clauses, cnf.vars).solve()
+}
+
+
+class CDCL(private var clauses: ArrayList<ArrayList<Int>>, private val varsNumber: Int) {
+    enum class VarStatus {
+        TRUE, FALSE, UNDEFINED;
+
+        operator fun not(): VarStatus {
+            return when (this) {
+                TRUE -> FALSE
+                FALSE -> TRUE
+                UNDEFINED -> UNDEFINED
+            }
+        }
+    }
+
+    private fun getStatus(lit: Int): VarStatus {
+        if (vars[litIndex(lit)].status == VarStatus.UNDEFINED) return VarStatus.UNDEFINED
+        if (lit < 0) return !vars[-lit].status
+        return vars[lit].status
+    }
+
+    private fun setStatus(lit: Int, status: VarStatus) {
+        if (lit < 0) {
+            vars[-lit].status = !status
+        } else {
+            vars[lit].status = status
+        }
+    }
+
+    data class VarState(
+        var status: VarStatus,
+        var clause: Int,
+        var level: Int,
+    )
+
+    // convert values to a possible satisfying result: if a variable less than 0 it's FALSE, otherwise it's TRUE
+    private fun variableValues() = vars
+        .mapIndexed { index, v ->
+            when (v.status) {
+                VarStatus.TRUE -> index
+                VarStatus.FALSE -> -index
+                else -> index
+            }
+        }.sortedBy { litIndex(it) }.filter { litIndex(it) > 0 }
+
+    // values of variables
+    private val vars: MutableList<VarState> = MutableList(varsNumber + 1) { VarState(VarStatus.UNDEFINED, -1, -1) }
+
+    // all decisions and consequences
+    private val trail: ArrayList<Int> = ArrayList()
+
+    // decision level
+    private var level: Int = 0
+
+    // two watched literals heuristic
+    private val watchers = MutableList(varsNumber + 1) { mutableSetOf<Int>() } // set of clauses watched by literal
+    private fun litIndex(lit: Int): Int = abs(lit)
+
+    // list of clauses to propagate
```

Comment:
This is not a "list of clauses", this is either a "list of literals" or "list of unit clauses" (or "list of units").
Also, it is advisable to explicitly specify types for public/important fields. This field is private, but important, so maybe make it `private val units: MutableList<Int> =...` in order to make it perfectly clear what type it has?

Reply:
OK